### PR TITLE
Highlight indent guide

### PR DESCRIPTION
**Pull request by mkslanc**
*Description of changes:*

New option to provide ability of indent guide highlighting for all ace themes

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/layer/text.js [hunk 1]
```diff
@@ -101,6 +101,15 @@ var Text = function(parentEl) {
         return true;
     };
 
+    this.highlightIndentGuides = true;
+    this.setHighlightIndentGuides = function (highlight) {
+        if (this.displayIndentGuides == false) return false;
```
Comment (andrewnester) [inline src/layer/text.js:106]: better to use strict equal here

File: src/layer/text.js [hunk 2]
```diff
@@ -101,6 +101,15 @@ var Text = function(parentEl) {
...
+        if (this.highlightIndentGuides == highlight) return false;
```
Comment (andrewnester) [inline src/layer/text.js:107]: better to use strict equal here

File: src/layer/text.js [hunk 3]
```diff
@@ -404,21 +414,131 @@ var Text = function(parentEl) {
...
+
+        this.$highlightIndentGuideMarker = {
+            indentLevel: undefined,
+            start: undefined,
+            end: undefined,
+            dir: undefined
+        };
+        var lines = this.session.doc.$lines;
+        var cursor = this.session.selection.getCursor();
+        var initialIndent = /^\s*/.exec(this.session.doc.getLine(cursor.row))[0].length;
```
Comment (andrewnester) [inline src/layer/text.js:441]: what if there are no spaces in the line? it seems like this line would throw exception

**Review by andrewnester — CHANGES_REQUESTED** [review]

File: src/layer/text.js (hunk 3 above)
Comment (mkslanc) [reply src/layer/text.js:441]: It would just return 0 in that case, as i am using `*` quantifier

**Comment by andrewnester** [comment]
One concern I have is the performance implication of this change - now it seems on every bracket render we'll go through all lines and all cells to render indent guides which can be quite expensive operation. @nightwing what is your thought on this?

File: src/layer/text.js (hunk 1 above)
Comment (mkslanc) [reply src/layer/text.js]: Done

File: src/layer/text.js (hunk 2 above)
Comment (mkslanc) [reply src/layer/text.js]: Done

File: src/layer/text.js [hunk 4]
```diff
@@ -101,6 +101,15 @@ var Text = function(parentEl) {
         return true;
     };
 
+    this.highlightIndentGuides = true;
```
Comment (nightwing) [inline src/layer/text.js:104]: other places use $ to mark variables like this as private `this.$highlightIndentGuides = true;`

File: src/layer/text.js [hunk 5]
```diff
@@ -404,21 +414,131 @@ var Text = function(parentEl) {
...
+    };
+
+    this.$clearActiveIndentGuide = function () {
+        var cells = this.$lines.cells;
+        for (var i = 0; i < cells.length; i++) {
+            var cell = cells[i];
+            var childNodes = cell.element.childNodes;
+            if (childNodes.length > 0) {
+                for (var j = 0; j < childNodes.length; j++) {
+                    if (childNodes[j].className && childNodes[j].className.search("ace_indent-guide-active") !== -1) {
```
[the remark this answers is not in the document]
Comment (mkslanc) [reply src/layer/text.js:489]: Yeah, you right using classlist methods is more performant and we should stop after finding first one, changing

File: src/layer/text.js [hunk 6]
```diff
@@ -404,21 +414,131 @@ var Text = function(parentEl) {
...
+            }
+        }
+    };
+
+    this.$setIndentGuideActive = function (cell, indentLevel) {
+        var line = this.session.doc.getLine(cell.row);
+        if (line !== "") {
+            var childNodes = cell.element.childNodes;
+            if (childNodes && childNodes[indentLevel - 1] && childNodes[indentLevel - 1].className) {
+                childNodes[indentLevel - 1].className = childNodes[indentLevel - 1].className.replace(
```
[the remark this answers is not in the document]
Comment (mkslanc) [reply src/layer/text.js:503]: For me it doesn't matter much

**Review by andrewnester — APPROVED** [review]

**Pull request merged**